File: src/infrastructure/xzdec/xzdec.c
#include <stdint.h>
#include <stdlib.h>

#include <lzma.h>

typedef uint32_t xzdec_lzma_ret;

struct Xzdec {
	lzma_stream stream;
};

__attribute__((export_name("xzdec_allocate")))
uint8_t *xzdec_allocate(uint32_t len) {
	return malloc(len);
}

__attribute__((export_name("xzdec_deallocate")))
void xzdec_deallocate(uint8_t *ptr) {
	free(ptr);
}

__attribute__((export_name("xzdec_new_stream_decoder")))
xzdec_lzma_ret xzdec_new_stream_decoder(
	uint32_t memlimit,
	uint32_t flags,
	struct Xzdec **xzdec_ptr
) {
	lzma_stream stream = LZMA_STREAM_INIT;
	lzma_ret rc = lzma_stream_decoder(&stream, memlimit, flags);
	if (rc != LZMA_OK) {
		return rc;
	}
	*xzdec_ptr = malloc(sizeof(struct Xzdec));
	(*xzdec_ptr)->stream = stream;
	return LZMA_OK;
}

__attribute__((export_name("xzdec_drop")))
void xzdec_drop(struct Xzdec *xzdec) {
	lzma_end(&(xzdec->stream));
	free(xzdec);
}

__attribute__((export_name("xzdec_input_empty")))
uint32_t xzdec_input_empty(struct Xzdec *xzdec) {
	if (xzdec->stream.avail_in == 0) {
		return 1;
	}
	return 0;
}

__attribute__((export_name("xzdec_set_input")))
void xzdec_set_input(
	struct Xzdec *xzdec,
	const uint8_t *input_buf,
	uint32_t input_buf_len
) {
	xzdec->stream.next_in = input_buf;
	xzdec->stream.avail_in = input_buf_len;
}

__attribute__((export_name("xzdec_next_output")))
xzdec_lzma_ret xzdec_next_output(
	struct Xzdec *xzdec,
	uint8_t *output_buf,
	uint32_t output_buf_cap,
	uint32_t *output_buf_len
) {
	xzdec->stream.next_out = output_buf;
	xzdec->stream.avail_out = output_buf_cap;
	lzma_ret rc = lzma_code(&(xzdec->stream), LZMA_RUN);
	*output_buf_len = output_buf_cap - xzdec->stream.avail_out;
	return rc;
}

__attribute__((export_name("xzdec_finish")))
xzdec_lzma_ret xzdec_finish(
	struct Xzdec *xzdec,
	uint8_t *output_buf,
	uint32_t output_buf_cap,
	uint32_t *output_buf_len
) {
	xzdec->stream.next_out = output_buf;
	xzdec->stream.avail_out = output_buf_cap;
	lzma_ret rc = lzma_code(&(xzdec->stream), LZMA_FINISH);
	*output_buf_len = output_buf_cap - xzdec->stream.avail_out;
	return rc;
}

// Prevent Clang from wrapping every inserted function and injecting calls
// to `__wasm_call_dtors()`.
void _initialize() {
	void __wasm_call_ctors();
	__wasm_call_ctors();
}

File: src/infrastructure/xzdec/xzdec.ts
import { promises as fs } from "node:fs";
import stream from "node:stream";
import zlib from "node:zlib";

const LZMA_CONCATENATED = 0x08;

type lzma_ret = number;

const lzma_ret = {
  OK                 : 0,
  STREAM_END         : 1,
  NO_CHECK           : 2,
  UNSUPPORTED_CHECK  : 3,
  GET_CHECK          : 4,
  MEM_ERROR          : 5,
  MEMLIMIT_ERROR     : 6,
  FORMAT_ERROR       : 7,
  OPTIONS_ERROR      : 8,
  DATA_ERROR         : 9,
  BUF_ERROR          : 10,
  PROG_ERROR         : 11,
  SEEK_NEEDED        : 12,
} as const;

const BUF_SIZE = 0x10000; // 64 KiB

// 128 MiB, large enough for archives encoded with `xz -9` plus some
// extra margin;
const MEM_LIMIT = 0x8000000;

type ptr = number;

interface xzdec_exports {
  xzdec_allocate(len: number): ptr;
  xzdec_deallocate(ptr: ptr): void;

  xzdec_new_stream_decoder(
    memlimit: number,
    flags: number,
    xzdec_ptr: ptr,
  ): lzma_ret;

  xzdec_drop(xzdec: ptr): lzma_ret;

  xzdec_input_empty(xzdec: ptr): number;

  xzdec_set_input(xzdec: ptr, input_buf: ptr, input_buf_len: number): void;

  xzdec_next_output(
    xzdec: ptr,
    output_buf: ptr,
    output_buf_cap: number,
    output_buf_len: ptr,
  ): lzma_ret;

  xzdec_finish(
    xzdec: ptr,
    output_buf: ptr,
    output_buf_cap: number,
    output_buf_len: ptr,
  ): lzma_ret;
};

let moduleOnce: Promise<WebAssembly.Module> = null;

async function loadXzdec(): Promise<WebAssembly.Module> {
  const wasmPath = "./infrastructure/xzdec/xzdec.wasm.gz";
  const wasmGzBytes = await fs.readFile(wasmPath);
  const wasmBytes = new Uint8Array(zlib.gunzipSync(wasmGzBytes));
  return await WebAssembly.compile(wasmBytes);
}

export async function decompress(r: stream.Readable, w: stream.Writable) {
  if (moduleOnce === null) {
    moduleOnce = loadXzdec();
  }

  const instance = await WebAssembly.instantiate(await moduleOnce, {});
  const mem: WebAssembly.Memory = instance.exports.memory as any;
  const {
    xzdec_allocate,
    // xzdec_deallocate,
    xzdec_new_stream_decoder,
    // xzdec_drop,
    xzdec_input_empty,
    xzdec_set_input,
    xzdec_next_output,
    xzdec_finish,
  } = (instance.exports as any) as xzdec_exports;

  const SCRATCH_SIZE = 8;
  const scratchPtr = xzdec_allocate(SCRATCH_SIZE) as number;
  const inputPtr = xzdec_allocate(BUF_SIZE) as number;
  const outputPtr = xzdec_allocate(BUF_SIZE) as number;
  if (scratchPtr == 0 || inputPtr === 0 || outputPtr === 0) {
    throw new Error("xzdec_allocate() failed");
  }

    // struct scratch {
    //     struct Xzdec *;
    //     uint8_t *output_buf_len;
    // }
    const xzdecPtr = scratchPtr;
    const outputLenPtr = scratchPtr + 4;
  
    const flags = LZMA_CONCATENATED;
    let rc = xzdec_new_stream_decoder(MEM_LIMIT, flags, scratchPtr);
    if (rc !== lzma_ret.OK) {
      throw new Error(`xzdec_new_stream_decoder() failed: lzma_ret(${rc})`);
    }
    const xzdec = peekU32(mem, xzdecPtr);

    for await (let chunk of r) {
      while (chunk.length > 0) {
        if (xzdec_input_empty(xzdec) === 1) {
          const input = chunk.subarray(0, BUF_SIZE);
          chunk = chunk.subarray(BUF_SIZE);
          new Uint8Array(mem.buffer, inputPtr, input.length).set(input);
          xzdec_set_input(xzdec, inputPtr, input.length);
        }

        while (xzdec_input_empty(xzdec) == 0) {
          let rc = xzdec_next_output(xzdec, outputPtr, BUF_SIZE, outputLenPtr);
          if (rc !== lzma_ret.OK) {
            throw new Error(`xzdec_next_output() failed: lzma_ret(${rc})`);
          }
          const outputLen = peekU32(mem, outputLenPtr);
          if (outputLen > 0) {
            await new Promise((resolve) => {
              if (!w.write(Buffer.from(mem.buffer, outputPtr, outputLen))) {
                w.once('drain', resolve)
              }
              else {
                resolve(null);
              }
            });
          }
        }
      }
    }
  
    rc = xzdec_finish(xzdec, outputPtr, BUF_SIZE, outputLenPtr);
    if (rc !== lzma_ret.OK) {
      if (rc !== lzma_ret.STREAM_END) {
        throw new Error(`xzdec_finish() failed: lzma_ret(${rc})`);
      }
      const outputLen = peekU32(mem, outputLenPtr);
      if (outputLen > 0) {
        w.write(Buffer.from(mem.buffer, outputPtr, outputLen));
      }
    }

    // Be lazy and let the entire module get garbage-collected, instead of
    // deallocating buffers.
}

function peekU32(mem: WebAssembly.Memory, addr: ptr): number {
  return new Uint32Array(mem.buffer, addr)[0];
}
